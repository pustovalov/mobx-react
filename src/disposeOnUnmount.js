import * as React from "react"
import { newSymbol } from "./utils/utils"

const storeKey = newSymbol("disposeOnUnmount")
const baseUnmountKey = newSymbol("originalOnUnmount")

function checkFunc(prop) {
    if (typeof prop !== "function") {
        throw new Error(
            "[mobx-react] disposeOnUnmount only works on functions such as disposers returned by reactions, autorun, etc."
        )
    }
}

function check(prop) {
    if (Array.isArray(prop)) {
        prop.map(checkFunc)
    } else {
        checkFunc(prop)
    }
}

function runDisposersOnWillUnmount() {
    if (this[baseUnmountKey]) this[baseUnmountKey]()
    if (!this[storeKey]) {
        // when disposeOnUnmount is only set to some instances of a component it will still patch the prototype
        return
    }
    this[storeKey].forEach(propKeyOrFunction => {
        const prop =
            typeof propKeyOrFunction === "string" ? this[propKeyOrFunction] : propKeyOrFunction
        if (prop !== undefined && prop !== null) {
            check(prop)
            if (Array.isArray(prop)) prop.map(f => f())
            else prop()
        }
    })
    this[storeKey] = []
}

export function disposeOnUnmount(target, propertyKeyOrFunctionOrArray) {
    if (Array.isArray(propertyKeyOrFunctionOrArray)) {
        return propertyKeyOrFunctionOrArray.map(fn => disposeOnUnmount(target, fn))
    }

    const c = Object.getPrototypeOf(target).constructor || Object.getPrototypeOf(target.constructor)
    const c2 = Object.getPrototypeOf(target.constructor)
    if (
        !(
            c === React.Component ||
            c === React.PureComponent ||
            c2 === React.Component ||
            c2 === React.PureComponent
        )
    ) {
        throw new Error(
            "[mobx-react] disposeOnUnmount only supports direct subclasses of React.Component or React.PureComponent."
        )
    }

    if (typeof propertyKeyOrFunctionOrArray !== "string") {
        check(propertyKeyOrFunctionOrArray)
    }

    // add property key / function we want run (disposed) to the store
    const componentWasAlreadyModified = !!target[storeKey]
    const store = target[storeKey] || (target[storeKey] = [])

    store.push(propertyKeyOrFunctionOrArray)

    // tweak the component class componentWillUnmount if not done already
    if (!componentWasAlreadyModified) {
        // make sure original definition is invoked
        if (target.componentWillUnmount) target[baseUnmountKey] = target.componentWillUnmount

        Object.defineProperty(target, "componentWillUnmount", {
            get() {
                return runDisposersOnWillUnmount
            },
            set(fn) {
                // this will happen if componentWillUnmount is being assigned after patching the prototype
                this[storeKey].push(fn)
                // assigning a new local value to componentWillUnmount would hide the super implementation...
                this[baseUnmountKey] = undefined
            },
            configurable: false,
            enumerable: false
        })
    }

    // return the disposer as is if invoked as a non decorator
    if (typeof propertyKeyOrFunctionOrArray !== "string") {
        return propertyKeyOrFunctionOrArray
    }
}
