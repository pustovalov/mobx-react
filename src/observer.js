import React, { Component, PureComponent, forwardRef } from "react"
import { createAtom, _allowStateChanges } from "mobx"
import {
    observer as observerLite,
    useStaticRendering as useStaticRenderingLite,
    Observer
} from "mobx-react-lite"

import { newSymbol, shallowEqual } from "./utils/utils"

let isUsingStaticRendering = false

const skipRenderKey = newSymbol("skipRender")
const isForcingUpdateKey = newSymbol("isForcingUpdate")

// Using react-is had some issues (and operates on elements, not on types), see #608 / #609
const ReactForwardRefSymbol =
    typeof forwardRef === "function" && forwardRef((_props, _ref) => {})["$$typeof"]

/**
 * Helper to set `prop` to `this` as non-enumerable (hidden prop)
 * @param target
 * @param prop
 * @param value
 */
function setHiddenProp(target, prop, value) {
    if (!Object.hasOwnProperty.call(target, prop)) {
        Object.defineProperty(target, prop, {
            enumerable: false,
            configurable: true,
            writable: true,
            value
        })
    } else {
        target[prop] = value
    }
}

export function useStaticRendering(useStaticRendering) {
    isUsingStaticRendering = useStaticRendering
    useStaticRenderingLite(useStaticRendering)
}

function observerSCU(nextProps, nextState) {
    if (isUsingStaticRendering) {
        console.warn(
            "[mobx-react] It seems that a re-rendering of a React component is triggered while in static (server-side) mode. Please make sure components are rendered only once server-side."
        )
    }
    // update on any state changes (as is the default)
    if (this.state !== nextState) {
        return true
    }
    // update if props are shallowly not equal, inspired by PureRenderMixin
    // we could return just 'false' here, and avoid the `skipRender` checks etc
    // however, it is nicer if lifecycle events are triggered like usually,
    // so we return true here if props are shallowly modified.
    return !shallowEqual(this.props, nextProps)
}

function makeObservableProp(target, propName) {
    const valueHolderKey = newSymbol(`reactProp_${propName}_valueHolder`)
    const atomHolderKey = newSymbol(`reactProp_${propName}_atomHolder`)
    function getAtom() {
        if (!this[atomHolderKey]) {
            setHiddenProp(this, atomHolderKey, createAtom("reactive " + propName))
        }
        return this[atomHolderKey]
    }
    Object.defineProperty(target, propName, {
        configurable: true,
        enumerable: true,
        get: function() {
            getAtom.call(this).reportObserved()
            return this[valueHolderKey]
        },
        set: function set(v) {
            if (!this[isForcingUpdateKey] && !shallowEqual(this[valueHolderKey], v)) {
                setHiddenProp(this, valueHolderKey, v)
                setHiddenProp(this, skipRenderKey, true)
                getAtom.call(this).reportChanged()
                setHiddenProp(this, skipRenderKey, false)
            } else {
                setHiddenProp(this, valueHolderKey, v)
            }
        }
    })
}

/**
 * Observer function / decorator
 */
export function observer(componentClass) {
    if (componentClass.isMobxInjector === true) {
        console.warn(
            "Mobx observer: You are trying to use 'observer' on a component that already has 'inject'. Please apply 'observer' before applying 'inject'"
        )
    }

    // Unwrap forward refs into `<Observer>` component
    // we need to unwrap the render, because it is the inner render that needs to be tracked,
    // not the ForwardRef HoC
    if (ReactForwardRefSymbol && componentClass["$$typeof"] === ReactForwardRefSymbol) {
        const baseRender = componentClass.render
        if (typeof baseRender !== "function")
            throw new Error("render property of ForwardRef was not a function")
        return forwardRef(function ObserverForwardRef() {
            return <Observer>{() => baseRender.apply(undefined, arguments)}</Observer>
        })
    }

    // Function component
    if (
        typeof componentClass === "function" &&
        (!componentClass.prototype || !componentClass.prototype.render) &&
        !componentClass.isReactClass &&
        !Component.isPrototypeOf(componentClass)
    ) {
        return observerLite(componentClass)
    }

    return makeClassComponentObserver(componentClass)
}

function makeClassComponentObserver(componentClass) {
    const target = componentClass.prototype || componentClass
    if (target.componentWillReact)
        throw new Error("The componentWillReact life-cycle event is no longer supported")
    if (componentClass.__proto__ !== PureComponent) {
        if (!target.shouldComponentUpdate) target.shouldComponentUpdate = observerSCU
        else if (target.shouldComponentUpdate !== observerSCU)
            throw new Error(
                "It is not allowed to use shouldComponentUpdate in observer based components."
            )
    }
    makeObservableProp(target, "props")
    makeObservableProp(target, "state")
    const baseRender = target.render

    target.render = function renderWrapper() {
        if (!this.baseRender) {
            // safe the closure, as it won't change!
            const bound = baseRender.bind(this)
            this.baseRender = () => bound()
        }
        return <Observer>{this.baseRender}</Observer>
    }
    return componentClass
}
