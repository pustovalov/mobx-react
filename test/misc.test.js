import React from "react"
import createClass from "create-react-class"
import ReactDOM from "react-dom"
import * as mobx from "mobx"
import { observer } from "../src"
import { createTestRoot, withConsole } from "./index"
import renderer, { act } from "react-test-renderer"

const mobxAdminProperty = mobx.$mobx || "$mobx"

const testRoot = createTestRoot()

test("issue mobx 405", () => {
    function ExampleState() {
        mobx.extendObservable(this, {
            name: "test",
            get greetings() {
                return "Hello my name is " + this.name
            }
        })
    }

    const ExampleView = observer(
        createClass({
            render() {
                return (
                    <div>
                        <input
                            type="text"
                            onChange={e => (this.props.exampleState.name = e.target.value)}
                            value={this.props.exampleState.name}
                        />
                        <span>{this.props.exampleState.greetings}</span>
                    </div>
                )
            }
        })
    )

    const exampleState = new ExampleState()
    const wrapper = renderer.create(<ExampleView exampleState={exampleState} />)
    expect(wrapper.toJSON()).toMatchInlineSnapshot(`
<div>
  <input
    onChange={[Function]}
    type="text"
    value="test"
  />
  <span>
    Hello my name is test
  </span>
</div>
`)
})

test("#85 Should handle state changing in constructors", done => {
    const a = mobx.observable.box(2)
    const Child = observer(
        createClass({
            displayName: "Child",
            getInitialState() {
                a.set(3) // one shouldn't do this!
                return {}
            },
            render: () => (
                <div>
                    child:
                    {a.get()} -{" "}
                </div>
            )
        })
    )
    const ParentWrapper = observer(function Parent() {
        return (
            <span>
                <Child />
                parent:
                {a.get()}
            </span>
        )
    })
    ReactDOM.render(<ParentWrapper />, testRoot, () => {
        expect(testRoot.getElementsByTagName("span")[0].textContent).toBe("child:3 - parent:2")
        a.set(5)
        setTimeout(() => {
            expect(testRoot.getElementsByTagName("span")[0].textContent).toBe("child:5 - parent:5")
            a.set(7)
            setTimeout(() => {
                expect(testRoot.getElementsByTagName("span")[0].textContent).toBe(
                    "child:7 - parent:7"
                )
                testRoot.parentNode.removeChild(testRoot)
                done()
            }, 10)
        }, 10)
    })
})

test("testIsComponentReactive", () => {
    const C = observer(() => null)
    const wrapper = renderer.create(<C />)
    const instance = wrapper.getInstance()

    // instance is something different then the rendering reaction!
    expect(mobx.isObservable(instance)).toBeFalsy()
})

test("Do not warn about custom shouldComponentUpdate when it is the one provided by ReactiveMixin", () => {
    expect(
        withConsole(() => {
            const A = observer(
                class A extends React.Component {
                    render() {
                        return null
                    }
                }
            )

            observer(
                class B extends A {
                    render() {
                        return null
                    }
                }
            )
        })
    ).toMatchSnapshot()
})
