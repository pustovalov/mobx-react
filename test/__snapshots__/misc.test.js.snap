// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Do not warn about custom shouldComponentUpdate when it is the one provided by ReactiveMixin 1`] = `
Object {
  "errors": Array [],
  "infos": Array [],
  "warnings": Array [],
}
`;
